src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -931,102 +1021,60 @@ public virtual ODataResource CreateResource(SelectExpandNode selectExpandNode, R
-                {
-                    foreach (KeyValuePair<string, object> annotation in clrAnnotations)
-                    {
-                        AddODataAnnotations(resource.InstanceAnnotations, resourceContext, annotation);
-                    }
-                }
-                                
-                foreach(ODataProperty property in resource.Properties)
-                {
-                    string propertyName = property.Name;
-
-                    if (property.InstanceAnnotations == null)
-                    {
-                        property.InstanceAnnotations = new List<ODataInstanceAnnotation>();
-                    }
-
-                    IDictionary<string, object> propertyAnnotations = instanceAnnotationContainer.GetPropertyAnnotations(propertyName);
-
-                    if (propertyAnnotations != null)
-                    {
-                        foreach (KeyValuePair<string, object> annotation in propertyAnnotations)
-                        {
-                            AddODataAnnotations(property.InstanceAnnotations, resourceContext, annotation);
-                        }
-                    }                    
-                } 
+                delta = resourceContext.ResourceInstance as IDelta;
             }
-        }
-
-        private void AddODataAnnotations(ICollection<ODataInstanceAnnotation> InstanceAnnotations, ResourceContext resourceContext, KeyValuePair<string, object> annotation)
-        {
-            ODataValue annotationValue = null;
 
-            if (annotation.Value != null)
+            if (delta != null)
             {
-                IEdmTypeReference edmTypeReference = resourceContext.SerializerContext.GetEdmType(annotation.Value,
-                                            annotation.Value.GetType());
+                if (instanceAnnotationInfo != null)
+                {
+                    delta.TryGetPropertyValue(instanceAnnotationInfo.Name, out instanceAnnotations);
 
-                ODataEdmTypeSerializer edmTypeSerializer = GetEdmTypeSerializer(edmTypeReference);
+                }
+
+                IDeltaSetItem deltaitem = resourceContext.ResourceInstance as IDeltaSetItem;
 
-                if (edmTypeSerializer != null)
+                if (deltaitem != null)
                 {
-                    annotationValue = edmTypeSerializer.CreateODataValue(annotation.Value, edmTypeReference, resourceContext.SerializerContext);
+                    transientAnnotations = deltaitem.TransientInstanceAnnotationContainer;
                 }
             }
             else
             {
-                annotationValue = new ODataNullValue();
-            }
-
-            if (annotationValue != null)
-            {
-                InstanceAnnotations.Add(new ODataInstanceAnnotation(annotation.Key, annotationValue));               
-            }            
-        }
+                if (instanceAnnotationInfo == null || structuredObject == null ||
```

Comment:
Why do we have `structuredObject == null` as a condition here yet when that condition is true the code in the if block will be a no-op since `edmEntityObject` will be null?
I think this should be:
```suggestion
                if (instanceAnnotationInfo == null || (structuredObject != null && !structuredObject.TryGetPropertyValue(instanceAnnotationInfo.Name, out instanceAnnotations)) || instanceAnnotations == null)
```